straxen/plugins/event_patternfit.py

```diff
@@ -0,0 +1,410 @@
+        result['s1_bottom_2llh'][cur_s1_bool] = np.sum(norm_llh_val, axis=1)
+
+            
+    def compute_s2_llhvalue(self, events, result):
+        for t_ in ['s2', 'alt_s2']:
+            result[t_+'_2llh'][:] = np.zeros(len(events))
+
+            # Selecting S2s for pattern fit calculation
+            # - must exist (index != -1)
+            # - must have total area larger minimal one
+            # - must have positive AFT
+            x,y = events[t_+'_x'], events[t_+'_y']
+            cur_s2_bool = (events[t_+'_area']>self.config['s2_min_reconstruction_area'])
+            cur_s2_bool &= (events[t_+'_index']!=-1)
+            cur_s2_bool &= (events["s2_area_fraction_top"]>0)
+            cur_s2_bool &= (x**2 + y**2) < self.config['max_r']**2            
+            # Making expectation patterns [ in PE ]
+            s2_map_effs = self.s2_pattern_map(np.array([x, y]).T)[cur_s2_bool, 0:self.config['n_top_pmts']]
+            s2_map_effs = s2_map_effs[:, self.pmtbool_top]
+            s2_top_area = (events["s2_area_fraction_top"]*events["s2_area"])[cur_s2_bool]        
+            s2_pattern  = s2_top_area[:, None]*s2_map_effs/np.sum(s2_map_effs, axis=1)[:,None]  
+            
+            # Getting pattern from data
+            s2_top_area_per_channel = events['s2_area_per_channel'][cur_s2_bool, 0:self.config['n_top_pmts']]
+            s2_top_area_per_channel = s2_top_area_per_channel[:, self.pmtbool_top]
+            
+            # Calculating LLH, this is shifted Poisson
+            # we get area expectation and we need to scale them to get
+            # photon expectation
+            norm_llh_val = (neg2llh_modpoisson(
+                                 mu    = s2_pattern/self.mean_sPhoton, 
+                                 areas = s2_top_area_per_channel, 
+                                 mean_sPhoton=self.mean_sPhoton)
+                                    - 
+                            neg2llh_modpoisson(
+                                 mu    = s2_top_area_per_channel/self.mean_sPhoton, 
+                                 areas = s2_top_area_per_channel, 
+                                 mean_sPhoton=self.mean_sPhoton)
+                           )
+            result[t_+'_2llh'][cur_s2_bool]=np.sum(norm_llh_val, axis=1)
+            if self.config['StorePerChannel']:
+                result[t_+'_pattern'][:] = 0.0
+                result[t_+'_2llh_per_channel'][:] = 0.0
+                store_patterns = np.zeros((s2_pattern.shape[0], self.config['n_top_pmts']) )
+                store_patterns[:, self.pmtbool_top] = s2_pattern
+                result[t_+'_pattern'][cur_s2_bool] = store_patterns#:s2_pattern[cur_s2_bool]
+                
+                store_2LLH_ch = np.zeros((norm_llh_val.shape[0], self.config['n_top_pmts']) )
+                store_2LLH_ch[:, self.pmtbool_top] = norm_llh_val
+                result[t_+'_2llh_per_channel'][cur_s2_bool] = store_2LLH_ch
```

Comment:
Would it possible to simplify this a bit and re-use the overlapping code between S1s and S2?

Reply:
I am not entirely sure whether making generalized version will make it much easier, given different shapes (only top, vs total vs bottom arrays) etc. And it cuts off the possibility of treating S1 and S2 differently, if we decide so. Probably it is easier to leave it as it is.